File: cpp/src/topo_sort.cpp
#include "topo_sort.h"

#include <algorithm>
#include <deque>
#include <iostream>

// Topological sorting using BFS (Kahn's algorithm).
// https://en.wikipedia.org/wiki/Topological_sorting#Kahn's_algorithm
//
// Time complexity: O(V + E)
// Space complexity: O(V)
//
// `edges` is a list of edges in the form `(from, to)` where `from` and `to` are
// node indices.
std::optional<std::vector<Node>> topo_sort_kahn(uint32_t number_of_nodes,
                                                AdjacencyMatrix edges) {
  std::vector<Node> result = {};
  result.reserve(number_of_nodes);

  auto degrees = std::vector<uint32_t>(number_of_nodes, 0);

  for (auto [from, to] : edges) {
    if (from != to) {
      degrees[to]++;
    }
  }

  std::deque<Node> queue{};

  for (auto i = 0; i < degrees.size(); i++) {
    if (degrees[i] == 0) {
      queue.push_back(i);
    }
    std::cout << i << ":" << degrees[i] << std::endl;
  }

  while (!queue.empty()) {
    std::cout << "Q: ";
    for (auto i = 0; i < queue.size(); i++) {
      std::cout << queue[i] << " ";
    }
    std::cout << std::endl;

    auto node = queue.front();
    queue.pop_front();
    result.push_back(node);
    // look up outgoing nodes
    for (auto [from, to] : edges) {
      if (from == node) {
        if (--degrees[to] == 0) {
          queue.push_back(to);
        }
      }
    }
  }

  if (result.size() != number_of_nodes) {
    return std::nullopt;
  }
  return result;
}

void print(std::vector<Node> result) {
  std::cout << "[";
  for (auto i = 0; i < result.size(); i++) {
    std::cout << result[i] << " ";
  }
  std::cout << "]" << std::endl;
}
bool dfs(std::vector<Node> &recursion_stack, Node node, AdjacencyMatrix edges,
         std::vector<bool> &visited, std::vector<Node> &result) {
  if (!visited[node]) {
    visited[node] = true;

    for (auto [from, to] : edges) {
      if (from == node) {
        // Check if there is a cycle.
        if (std::find(recursion_stack.begin(), recursion_stack.end(), to) !=
            recursion_stack.end()) {
          return false;
        }
        recursion_stack.push_back(to);

        if (!dfs(recursion_stack, to, edges, visited, result)) {
          // Propagate detected cycle;
          return false;
        }

        recursion_stack.pop_back();
      }
    }
    result.push_back(node);
  }
  return true;
}

std::optional<std::vector<Node>> topo_sort_dfs(uint32_t number_of_nodes,
                                               AdjacencyMatrix edges) {
  std::vector<Node> result = {};
  std::vector<Node> recursion_stack = {};

  std::vector<bool> visited(number_of_nodes, false);

  for (auto i = 0; i < number_of_nodes; i++) {
    if (!visited[i]) {
      recursion_stack.push_back(i);
      if (!dfs(recursion_stack, i, edges, visited, result)) {
        return std::nullopt;
      }
      recursion_stack.pop_back();
    }
  }
  std::reverse(result.begin(), result.end());
  return result;
}

File: rust/src/topo_sort.rs
use std::collections::HashSet;

pub type Node = usize;

pub fn topo_sort_kahn(number_of_nodes: usize, edges: Vec<(Node, Node)>) -> Option<Vec<Node>> {
    let mut result = Vec::with_capacity(number_of_nodes);

    if number_of_nodes == 0 {
        return result.into();
    }

    let mut in_degree = vec![0; number_of_nodes];

    for &(_, to) in &edges {
        in_degree[to] += 1;
    }

    let mut queue = Vec::with_capacity(number_of_nodes);
    for i in 0..number_of_nodes {
        if in_degree[i] == 0 {
            queue.push(i);
        }
    }

    while !queue.is_empty() {
        let node = queue.pop().unwrap();
        result.push(node);

        for to in edges
            .iter()
            .filter(|(from, _)| &node == from)
            .map(|&(_, to)| to)
        {
            in_degree[to] -= 1;
            if in_degree[to] == 0 {
                queue.push(to);
            }
        }
    }
    (result.len() == number_of_nodes).then_some(result)
}

#[derive(Debug)]
pub struct CycleDetected(Node, Node);
impl std::fmt::Display for CycleDetected {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "cycle detected: {} <-> {}", self.0, self.1)
    }
}

pub fn topo_sort_dfs(
    number_of_nodes: usize,
    edges: Vec<(Node, Node)>,
) -> Result<Vec<Node>, CycleDetected> {
    let mut result = Vec::with_capacity(number_of_nodes);
    let mut visited = vec![false; number_of_nodes];
    let mut stack = HashSet::with_capacity(number_of_nodes);

    for i in 0..number_of_nodes {
        if !visited[i] {
            stack.insert(i);
            dfs(i, &edges, &mut visited, &mut stack, &mut result)?;
            stack.remove(&i);
        }
    }

    result.reverse();
    Ok(result)
}

fn dfs(
    node: Node,
    edges: &[(Node, Node)],
    visited: &mut [bool],
    stack: &mut HashSet<Node>,
    result: &mut Vec<Node>,
) -> Result<(), CycleDetected> {
    if !visited[node] {
        visited[node] = true;

        for to in edges
            .iter()
            .filter(|(from, _)| &node == from)
            .map(|&(_, to)| to)
        {
            if !stack.insert(to) {
                // If a node is already in the stack, it means that there is a cycle.
                return Err(CycleDetected(node, to));
            }
            dfs(to, edges, visited, stack, result)?;
            stack.remove(&to);
        }
        result.push(node);
    }

    Ok(())
}
